Src/Support/Google.Apis.Auth/OAuth2/TokenRefreshManager.cs

```diff
@@ -0,0 +1,170 @@
+﻿/*
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Logging;
+using Google.Apis.Util;
+using System;
+using System.Linq;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>
+    /// Encapsulation of token refresh behaviour. This isn't entirely how we'd design the code now (in terms of the
+    /// callback in particular) but it fits in with the exposed API surface of ServiceCredential and UserCredential.
+    /// </summary>
+    internal sealed class TokenRefreshManager
+    {
+        // Immutable state
+        private readonly object _lock = new object();
+        private readonly IClock _clock;
+        private readonly ILogger _logger;
+        private readonly Func<CancellationToken, Task<bool>> _refreshAction;
+
+        // Mutable state, guarded with _lock.
+        private TokenResponse _token;
+        private Task<TokenResponse> _refreshTask;
+
+        /// <summary>
+        /// Creates a manager which executes the given refresh action when required.
+        /// </summary>
+        /// <param name="refreshAction">The refresh action which will populate the Token property when successful.</param>
+        /// <param name="clock">The clock to consult for timeouts.</param>
+        /// <param name="logger">The logger to use to record refreshes.</param>
+        internal TokenRefreshManager(Func<CancellationToken, Task<bool>> refreshAction, IClock clock, ILogger logger)
+        {
+            _refreshAction = refreshAction;
+            _clock = clock;
+            _logger = logger;
+        }
+
+        internal TokenResponse Token
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    return _token;
+                }
+            }
+            // The token may be set due to operations other than GetAccessTokenForRequestAsync, but we don't need to
+            // null out _refreshTask if so; it will be 
+            set
+            {
+                lock (_lock)
+                {
+                    _token = value;
+                }
+            }
+        }
+
+        internal async Task<string> GetAccessTokenForRequestAsync(string authUri, CancellationToken cancellationToken)
```

Comment:
authUri parameter isn't used. Remove?

Reply:
Hmm... maybe. It's there to match the signature used in the original code, so it'll mean that that parameter isn't used instead, but that's reasonable. Will change.